### PR TITLE
Add `mixnode` and `mixnet-client` crate

### PR DESCRIPTION
**Pull request by youngjoon-lee**
Implemented the `mixnode` and `mixnet-client` crate, according to https://www.notion.so/Mixnet-Architecture-613f53cf11a245098c50af6b191d31d2.

NOTE:
- This PR is going to be merged to the `mixnet` branch, not the `master`.
- I recommend you to look at the test files first.

CHANGES:
- `mixnode`: A core crate for implementing the mixnode binary
  - forwards Sphinx packets to next mixnodes. 
- `mixnet-client`: A client library for mixnet users (such as nomos-node) 
  - splits a message into multiple Sphinx packets
  - sends Sphinx packets to the `mixnode`s.
  - receives Sphinx packets from `mixnodes`s and re-assemble the message.

TODO:
- Implement a `nomos-mixnode` binary using the `mixnode` crate.
- Integrate the `mixnet_client` crate with Nomos network backends.
- Refactoring
- Many TODOs written in source code.
- Cover traffic (We could postpone this a bit, as it's complex).

### CODE REVIEW

File: mixnet/client/src/sender.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            let packet = self.build_sphinx_packet(fragment, &destination, num_hops)?;
+            packets.push(packet);
+        }
+
+        for (packet, first_node) in packets {
+            tokio::spawn(async move {
+                if let Err(e) = Sender::send_packet(Box::new(packet), first_node.address).await {
+                    tracing::error!("failed to send packet to the first node: {e}");
+                }
+            });
```
Comment (Zeegomo) [inline mixnet/client/src/sender.rs:49]: nit: `Sender::pad_and_split_message(msg).map(|..| build_sphinx_packet(..)).collect::<Result<_>>()?` (yes you can go from an iterator of results into a result of a collection) and maybe even `.for_each(tokio::spawn(..))`

File: mixnet/mixnode/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        tokio::time::sleep(delay.to_duration()).await;
+
+        let addr = SocketAddr::try_from(NymNodeRoutingAddress::try_from(next_node_addr)?)?;
+        tracing::debug!("Forwarding a Sphinx packet to the mixnode: {addr:?}");
+
+        let mut socket = TcpStream::connect(addr).await?;
+        socket.write_all(&packet.to_bytes()).await?;
+
+        Ok(())
+    }
```
Comment (Zeegomo) [inline mixnet/mixnode/src/lib.rs:92]: this can reuse `forward_payload_to_destination`

**Review by Zeegomo — CHANGES_REQUESTED** [review]

File: mixnet/client/src/receiver.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        Ok(())
+    }
+
+    async fn handle_connection(
+        mut socket: TcpStream,
+        message_tx: broadcast::Sender<Vec<u8>>,
+        message_reconstructor: Arc<Mutex<MessageReconstructor>>,
+    ) -> Result<(), Box<dyn Error>> {
+        let mut buf = Vec::new();
+        socket.read_to_end(&mut buf).await?;
```
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply mixnet/client/src/receiver.rs:61]: Yes, but I think we need to introduce better connection managements at some point, depending on how often the mixnet topology is updated.

File: mixnet/client/src/receiver.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+        let payload = Payload::from_bytes(&buf)?.recover_plaintext()?;
+        let fragment = Fragment::try_from_bytes(&payload)?;
```
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply mixnet/client/src/receiver.rs:64]: If I understand your comment correctly, this function just receives a fragment (chunk) and throws it to the message reconstructor. The message reconstructor keeps all fragments previously received, in memory. If the message can be finally reconstructed with the newly received fragment, the message reconstrcutor returns the message fully reconstructed.

File: mixnet/client/src/sender.rs (hunk 1 above)
Comment (youngjoon-lee) [reply mixnet/client/src/sender.rs]: thank you: 20be7cf

File: mixnet/mixnode/src/lib.rs (hunk 2 above)
Comment (youngjoon-lee) [reply mixnet/mixnode/src/lib.rs]: Good point: 9b13f44e8cfc76d1d98358ed20b2e9df2ea2dd59

File: mixnet/topology/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                    .unwrap(),
+            );
+        }
+
+        Ok(route)
+    }
+}
+
+impl Layer {
+    pub fn random_node(&self) -> Option<&Mixnode> {
```
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply mixnet/topology/src/lib.rs]: Good point: 2fd055b.
I had to downgrade the `rand` and `rand_xoshiro` in the integration test, because `nym-sphinx` depends on the `rand@v0.7.3`. If we use the `rand@v0.8`, the following error occurs:
```rust
   --> mixnet/client/src/sender.rs:65:35
    |
65  |             .split_into_fragments(rng, plaintext_size_per_packet)
    |              -------------------- ^^^ the trait `rand_core::RngCore` is not implemented for `R`
```

File: mixnet/topology/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    pub fn random_route(&self, num_hops: usize) -> Result<Vec<route::Node>, Box<dyn Error>> {
+        if self.layers.len() < num_hops {
+            todo!("return error");
+        }
+
+        let mut route: Vec<route::Node> = Vec::new();
+
+        for layer_id in 0..num_hops {
+            let layer = self.layers.get(layer_id).unwrap();
```
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply mixnet/topology/src/lib.rs]: Thank you. I just realized that we don't actually need the `LayerId` type.

File: mixnet/node/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+use sphinx_packet::crypto::PRIVATE_KEY_SIZE;
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct MixnetNodeConfig {
+    // A listen address for receiving Sphinx packets
+    pub listen_address: SocketAddr,
+    // A key for decrypting Sphinx packets
+    pub private_key: [u8; PRIVATE_KEY_SIZE],
+    // An address of the client with receiver-mode enabled, for sending final payloads
+    pub client_address: SocketAddr,
```
Comment (youngjoon-lee) [inline mixnet/node/src/config.rs:14]: @Zeegomo @danielSanchezQ I just added 80e49f2 that changes the design a bit. Sorry to put this commit in this PR, but I think it's even easier to review.

In the previous design, the problem was that message senders must specify the mixnet destination (== the IP of `MixnetClient`).
Please note that it's different from the message routing destination).
But, in reality, senders doesn't have any information about the IP of `MixnetClient`s available.

Instead, I refined the design like this:
- A `nomos-node` always embeds a `MixnetClient` that just receives Sphinx packets, reconstructs messages, and sends them to its network backend.
   - Actually, this is the same in the previous design.
- If you want to run a `MixnetNode`, you must know at least one `MixnetClient` address.
  - In other words, `MixnetNode` operators should run their own `nomos-node` as well.
- When message senders choose 3 `MixnetNode` from `MixnetTopology`, they choose one more `MixnetNode` as a destination, randomly.
- Finallly, all message fragments (sent from a message sender) are forwarded to the destination `MixnetNode`. Then, the message is reconstructed and is broadcasted/routed to all other `nomos-node`s.
- On the other hand, as a `nomos-node` operator, it's not mandatory to run `MixnetNode`.

image

With this design, the message sender doesn't need to know about the mixnet destination.

File: tests/src/tests/mixnet.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+#[tokio::test]
+async fn mixnet() {
+    let topology = run_mixnodes();
+    let (client1, client2, destination) = run_clients(topology.clone()).await;
+
+    let mut msg = [0u8; 100 * 1024];
+    rand::thread_rng().fill_bytes(&mut msg);
+
+    let res = client1.send(msg.to_vec(), destination, &mut OsRng, topology.layers.len());
```
Comment (youngjoon-lee) [inline tests/src/tests/mixnet.rs:19]: I just changed the mixnet design a bit in the PR #307, so that the message sender doesn't need to specify the mixnet destination explicitly. Please review that PR together with this PR. I put detailed descriptions there.

File: mixnet/topology/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply mixnet/topology/src/lib.rs]: I also simplified the API by removing the `num_hops`: b8d3bad. We can revive it later after the first iteration is done.

**Review by danielSanchezQ — APPROVED** [review]
We can go ahead with this. 🚀

**Comment by youngjoon-lee** [comment]
@Zeegomo I'm merging this PR since I've applied your suggestions and there have been many architectural changes (especially in PR #307). I think it's better to merge this now to the `mixnet` branch, so that we can improve it gradually together.
